irc/client.go

```diff
@@ -1227,6 +1227,9 @@ func (client *Client) destroy(session *Session) {
 	} else {
 		sessionRemoved, remainingSessions = client.removeSession(session)
 		if sessionRemoved {
+			if alwaysOn || remainingSessions > 0 {
```

Comment:
Sorry, I didn't look closely enough before. You shouldn't send this here because you're still holding the client lock.

This can be moved down to the loop starting at L1285 ("destroy all applicable sessions"). The bottom of the loop looks like an appropriate place. You can just test `!shouldDestroy` in order to avoid duplicating these messages with the QUIT sno.

Reply:
```
if !shouldDestory {
 client.server.snomasks.Send(sno.LocalDisconnects, fmt.Sprintf(ircfmt.Unescape("Client session disconnected for [a:%s] [h:%s] [ip:%s]"), details.accountName, session.rawHostname, session.IP().String()))
			}

Reply:
This isn't what you meant though yeah?
Basically in the `for` loop you're handling  session disconnects. So right at the end before the logger. I can send the Sno?
Starting at L1310
```go
			client.server.connectionLimiter.RemoveClient(flatip.FromNetIP(ip))
			source = ip.String()
		}
		client.server.snomasks.Send(sno.LocalDisconnects, fmt.Sprintf(ircfmt.Unescape("Client session disconnected for [a:%s] [h:%s] [ip:%s]"), details.accountName, session.rawHostname, session.IP().String()))
		client.server.logger.Info("connect-ip", fmt.Sprintf("disconnecting session of %s from %s", details.nick, source))
	}
```
What I don't understand however is your suggestion about using `!shouldDestroy` to check that it doesn't send both `QUIT` and `DISCONNECT`. What is the expected logic there. `shouldDestory` is basically saying that there is no reason to send anything but QUIT if `true` yes?


Also im using `session.Rawhostnam` and `session.IP()String()` instead of `source` is that fine?